### ISSUE TITLE
(in-progress) Network stops working under load.

### ISSUE THREAD

**Issue by koshkamau** [morrownr/8821cu#16]
Hi!
I have a problem which (I think) is not related to this particular driver, but maybe I could get a piece of advice here anyway :)

I'm trying to use this driver (for USB-WiFi) on Allwinner A10 device (motherboard from old tablet). 
The problem is - driver stops working under load. 
In case of this NIC I use iperf3 to generate traffic. I have average bandwidth about 150Mbit (230Mbit MAX, up to 60% CPU load).
I just run iperf3 to my router and usually after a few lines networking stops.
There is no any error messages in syslog (messages were enabled). It's just like nothing happened.

I usually can revive adapter by executing
rmmod 8821cu
modprobe 8821cu
but not always.

I do not use DKMS (I don't know how). I just compile the driver against source code of the kernel that I use.
Currently I use 4.19.190 kernel. I gradually downgraded from 5.12 in hope that older kernel is better for WiFi.
Now I use 3.3V power supply for adapter but it has the same behavior with 5V supply.
I tweaked TCP/IP for better network performance.

I spent a couple of months investigating network problems with my devices but I still have no clue about what's happening.
Originally I had similar problem with usb-ethernet adapter and I even filed a bug-report to mainline kernel:
https://bugzilla.kernel.org/show_bug.cgi?id=212731
After I failed with ASIX I bought 8821CU adapter and get network disappearance too.

One more thing:
sometimes adapter works significantly longer before fails. I try to determine circumstances in which it has happened but usually I can't reproduce them.
It seemed to me that enabled logging makes driver (and network) slower so it works without errors for a very long time, but I couldn't reproduce it after reboot.
It also seems to me, that reloaded (rmmod/modprobe) driver works much better. 
But maybe I just see random values and try to make sense of them...

Can you give me a hint - what is happening with my system?
Did I miss something in kernel config?
Should I use DKMS?

Thanks!

**Comment by morrownr** [comment]
Hi koshkamau,

Are you running in AP mode?

**Comment by koshkamau** [comment]
No. Client. 
wpa_supplicant + systemd-networkd

**Comment by morrownr** [comment]
It appears the Allwinner A10 is an ARM SoC. What operating system are you running?

For what it is worth: This is actually a pretty high traffic site. This is a very stable driver in managed (client) mode. I don't think this chipset is popular for AP or monitor mode or we would see bug reports because I do test those modes and I know the bugs are there.

The only ARM platform I test on is Raspberry Pi but I'll try to help if I can.

**Comment by koshkamau** [comment]
Thank you.
I'm using Ubuntu 20.04 rootfs made by debootstrap so it may lack something.

**Comment by morrownr** [comment]
Are you using any module parameters?

**Comment by koshkamau** [comment]
I manually copied 8821cu.conf file to /etc/modprobe.d
I used only parameters that were in file.

I'm intended to test this adapter tonight on an ordinary hardware - my x86 laptop.

**Comment by morrownr** [comment]
Testing on a x86 system is a good idea, Try to use the installation instructions to include using the install-driver.sh script so as to test the full installation experience that has been tested here. If that proves to be successful then we have to look at what the issue really is but that could be hard to find. One thing just popped into my mind:

If it is in your budget, you could look at getting an adapter with in-kernel drivers as that would/could eliminate some possible issues:

https://github.com/morrownr/USB-WiFi

**Comment by koshkamau** [comment]
I've already read your guide on buying linux-compatible adapter today and searched on Aliexpress to see what they offer :)
Perhaps I will do that though I don't think it guarantees me anything.
As I mentioned in the first post I began with the safest strategy I could imagine: I bought **wired** adapter which has **in-kernel** drivers plus source code from the manufacturer.
And it doesn't work reliably even on my laptop (on Linux).

**Comment by morrownr** [comment]
You are correct that buying an adapter with good in-kernel support does not guarantee you anything. However, it is an option to consider as you move forward and things become more clear. I've been using Linux since the mid-90's and something that I discovered along the way is that what one might think is the safest strategy is maybe not the safest strategy. You kind of have to get a good feel for things and be an excellent investigator... and things can change over time.

What I do know is that the multitude of ARM SoCs all seem to have various little differences that have to be accounted for. Since I don't have an Allwinner A10, I am no expert on it. Have you been searching the internet for other reports of similar issues? Try various searches from narrow to wide as the cause could manifest in similar but different ways. Maybe somebody has already been down this road.

**Comment by koshkamau2** [comment]
It's me, koshkamau. My network provider changed terms of service and withdrew my public IP address tonight so my email server doesn't work and I can't get verification code to my previous account. I temporarily registered a new one.

I made a test on KDE Neon tonight and found that your driver is quite stable on my x86 laptop. I didn't wait for too long though - just about 1600 lines of iperf3 on the screen.

There is a difference in iperf behavior between Allwinner and laptop:

When I run iperf on Allwinner I always have traffic. All lines, which iperf prints on the screen have non-zero speed value. At least 40Mbps. But then it suddenly stops printing lines and never resumes until network is restarted.

When I run iperf on my laptop - iperf never stops printing lines, but there are gaps in traffic. I've seen up to 30 lines in a row with speed 0.00 Mbps. 

I don't know what's happening but I would suggest (what else can I do) that driver is actually falls or freezes in both cases, but when it is loaded using DKMS it's able to restart itself and when it is simply compiled against the kernel tree it somehow loses this ability?

Now I'm trying to figure out how to make dkms module...
I have no headers to build it on the target and I don't know how to cross-compile it...

P.S.
If you are asking about ASIX device then yes. I've been searching for a long time and found a few complains like "asix keeps disconnecting". There were complains about other chipsets too but they were all unanswered.

Anyway, current problems apart, there is no way I would return to ASIX. 
The WiFi module I have costs about 7$. It's cheaper than ASIX. 
It's 10 times or more smaller than ASIX
Its driver (thanks to you) have much higher throughput than ASIX on my hardware (USB2.0)
I had steady 110Mbps with ASIX on modern kernels. I have about 240MBps with WiFi on 5V power supply.

**Comment by morrownr** [comment]
I'm going to setup for testing this driver to see if I can duplicate what you are seeing later today.

I'd be interested to know how your wifi access point is configured. Do you run both 2.4g and 5g? Do they both use the same name?

**Comment by morrownr** [comment]
Test results with this driver:

Adapter: rtl8811cu based

OS: Linux Mint 20.1 (kernel 5.4)

Router: RasPi4b, RasPiOS, Alfa AWUS036ACM adapter, 5 Ghz 

$ iperf3 -c 192.168.1.41 -t 300

[ ID] Interval                 Transfer          Bitrate               Retr
[  5]   0.00-300.00 sec  8.33 GBytes   238 Mbits/sec    0             sender
[  5]   0.00-300.02 sec  8.32 GBytes   238 Mbits/sec                  receiver

No problems noted.

**Comment by koshkamau2** [comment]
Actually, 300sec maybe not enough. Yesterday, before I began laptop test I had made 10 runs on Allwinner.
I just started iperf, waited for it to die, recorded number of lines it had produced, rebooted system.
I made this 10 times and here is a table:
```
Attempt N     Lines
1               178
2               461
3               256
4               107
5               1326
6               690
7               626
8               655
9               462
10              67
```
One line is about one second, 28MB.

My router is Xiaomi Router 3G, AP runs both 2.4GHz and 5GHz using different SSIDs, OS is OpenWrt 19.07.4

**Comment by koshkamau** [comment]
There are options in Makefile:
```
CONFIG_PLATFORM_ARM_SUNxI = n
CONFIG_PLATFORM_ARM_SUN6I = n
CONFIG_PLATFORM_ARM_SUN7I = n
CONFIG_PLATFORM_ARM_SUN8I_W3P1 = n
CONFIG_PLATFORM_ARM_SUN8I_W5P1 = n
```

My platform is SUN5I so I think I should enable **CONFIG_PLATFORM_ARM_SUNxI**, but it gives error:
```
/8821cu/platform/platform_ARM_SUNxI_usb.c:23:10: fatal error: mach/sys_config.h: No such file or directory
   23 | #include <mach/sys_config.h>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by koshkamau** [comment]
I think this issue can be closed.
Sorry for all this disturbance :)
It seems that my board is faulty.
I did test on the other Allwinner A10 board and driver/adapter work fine there. 
I didn't notice this because the second board has its own problems with power supply for external drive which also lead to instability...

**Comment by morrownr** [comment]
No problem. I understand about problems that are hard to track down.

**Comment by koshkamau** [comment]
In case someone will come here from google or whatever...
As I understand the situation is slightly more complex than a faulty board...
In my oversimplified understanding, Allwinner A10 has 2 types of USB controllers.
The first type is USB-OTG which is Inventra Highspeed Dual Role Controller IP Core.
The second type - all other usb hosts if they are present.

I have two A10 based boards. Both, obviously, have USB-OTG. One board has an additional USB2.0 port, the other board has two additional USB2.0 ports.
I tested both boards and the error which I described here can be seen only on non-OTG USB2.0 ports.
USB-OTG (which is working in Host-only mode) works without errors though slower.
Currently I've been running this setup for several days without an issue.

So, on my Allwinner A10 boards (kernel 5.12):
1) 8821CU adapter has iperf speed up to 260Mbps if connected to non-OTG port, but it stops working suddenly.
2) 8821CU adapter has iperf speed up to 200Mbps if connected to USB-OTG port and works quite reliably.
3) I didn't test it by iperf but in actual applications WiFi is somewhat slower when transmitting data.
I mean uploading data on the server is faster than downloading from it, if 8821CU is installed on the server.
4) Actual receive speeds when using SAMBA: 
    Average 12MB/sec with spikes up to 17MB/sec when on non-OTG port
    Average 7MB/sec with spikes up to 11MB/sec when on USB-OTG port
5) Interesting thing - UBB-HDD works on non-OTG port without errors.
6) Also interesting - all these problems seems to be unrelated to my problems with ASIX. ASIX fails on both types of ports, it fails even on laptop.
ASIX can't be brought down by iperf, I also couldn't do it with SAMBA file transfers. ASIX dies only when Transmission is running...
